odoo/tools/sql.py

```diff
@@ -102,16 +102,17 @@ def convert_column(cr, tablename, columnname, columntype):
     _schema.debug("Table %r: column %r changed to type %s", tablename, columnname, columntype)
 
 def set_not_null(cr, tablename, columnname):
-    """ Add a NOT NULL constraint on the given column. """
+    """ Add a NOT NULL constraint on the given column, and return ``None`` (in
+        case of success) or an exception (in case of failure).
+    """
     query = 'ALTER TABLE "{}" ALTER COLUMN "{}" SET NOT NULL'.format(tablename, columnname)
     try:
         with cr.savepoint():
-            cr.execute(query)
+            cr.execute(query, log_exceptions=False)
             _schema.debug("Table %r: column %r: added constraint NOT NULL", tablename, columnname)
-    except Exception:
-        msg = "Table %r: unable to set NOT NULL on column %r!\n" \
-              "If you want to have it, you should update the records and execute manually:\n%s"
-        _schema.warning(msg, tablename, columnname, query, exc_info=True)
+    except Exception as e:
+        _schema.debug("Table %r: column %r: unable to set constraint NOT NULL", tablename, columnname)
+        return e
```

Comment:
We might want to serialize the traceback immediately, or at least use `traceback.clear_frames` in order to avoid reference cycles (see https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/)

Reply:
I never realized this could be an issue, but the linked article is enlightening.  Your suggestion is definitively worth giving it a try.  And it would avoid `exc_info=exc` issue in Python 2.